server/lib/chat/chat-models.ts

```diff
@@ -247,80 +237,96 @@ export async function getMessagesForChannel(
       SELECT name FROM channels
-      WHERE name = ${channelName} AND high_traffic = true`)
-    if (result.rowCount > 0) {
+      WHERE name = ${channelName} AND high_traffic = true;
+    `)
+    if (highTrafficChannelResult.rowCount > 0) {
       // Don't transfer ownership in "high traffic" channels
-      return { newOwnerId: null }
+      return { newOwnerId: undefined } as LeaveChannelResult
     }
 
-    result = await client.query<DbUserChannelEntry>(sql`
+    const earliestPermissionUserResult = await client.query<DbUserChannelEntry>(sql`
       SELECT *
       FROM channel_users
       WHERE channel_name = ${channelName} AND (kick = true OR ban = true OR
         change_topic = true OR toggle_private = true OR edit_permissions = true)
-      ORDER BY join_date`)
-    if (result.rowCount > 0) {
+      ORDER BY join_date;
+    `)
+    if (earliestPermissionUserResult.rowCount > 0) {
       // Transfer ownership to the user who has joined the channel earliest and has an
       // `edit_permissions` permission, or if there's no such user, then choose the first user with
       // any kind of permission
-      const newOwner = result.rows.find(u => u.edit_permissions) || result.rows[0]
+      const newOwner =
+        earliestPermissionUserResult.rows.find(u => u.edit_permissions) ||
+        earliestPermissionUserResult.rows[0]
+
       await client.query(sql`
-        UPDATE channel_users
-        SET kick = true, ban = true, change_topic = true, toggle_private = true,
-          edit_permissions = true, owner = true
-        WHERE user_id = ${newOwner.user_id} AND channel_name = ${channelName}`)
-      return { newOwnerId: newOwner.user_id }
+        UPDATE channels
+        SET owner_id = ${newOwner.user_id}
+        WHERE name = ${channelName};
+      `)
+      return { newOwnerId: newOwner.user_id } as LeaveChannelResult
     }
 
     // Transfer ownership to the user who has joined the channel earliest
-    result = await client.query(sql`
-      SELECT user_id, join_date
+    const earliestUserResult = await client.query<{ user_id: SbUserId }>(sql`
+      SELECT user_id
       FROM channel_users
       WHERE channel_name = ${channelName}
-      ORDER BY join_date`)
+      ORDER BY join_date;
+    `)
+
+    // This would mean that the channel has no users left at all which would be very odd indeed
+    if (earliestUserResult.rowCount < 1) {
+      throw new Error('No rows returned')
+    }
 
     await client.query(sql`
-      UPDATE channel_users
-      SET kick = true, ban = true, change_topic = true, toggle_private = true,
-        edit_permissions = true, owner = true
-      WHERE user_id = ${result.rows[0].user_id} AND channel_name = ${channelName}`)
-    return { newOwnerId: result.rows[0].user_id }
+      UPDATE channels
+      SET owner_id = ${earliestUserResult.rows[0].user_id}
+      WHERE name = ${channelName};
+    `)
+    return { newOwnerId: earliestUserResult.rows[0].user_id } as LeaveChannelResult
```

Comment:
Very much don't understand why this would need a cast, is transact just not inferring types properly here? If so we should fix that instead of doing these casts on every return